pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/SegmentCommitterFactory.java

```diff
@@ -69,9 +70,13 @@ public SegmentCommitter createSegmentCommitter(boolean isSplitCommit, SegmentCom
 
     boolean uploadToFs = _streamConfig.isServerUploadToDeepStore();
     String peerSegmentDownloadScheme = _tableConfig.getValidationConfig().getPeerSegmentDownloadScheme();
-    // TODO: exists for backwards compatibility. remove peerDownloadScheme non-null check once users have migrated
-    if (uploadToFs || peerSegmentDownloadScheme != null) {
-      segmentUploader = new PinotFSSegmentUploader(_indexLoadingConfig.getSegmentStoreURI(),
+    String segmentStoreUri = _indexLoadingConfig.getSegmentStoreURI();
+
+    // We seem to allow the server instance to come up without a valid segment store uri. Hence, this check is needed.
+
+    if (peerSegmentDownloadScheme != null || (!StringUtils.isEmpty(segmentStoreUri) && uploadToFs)) {
```

Comment:
We don't need this extra check because we already checked at table creation time